### ISSUE TITLE
Key Vault Secret purge happens too soon after delete causes Status=409 Code="Conflict" Message="Secret is currently being deleted."

### ISSUE THREAD

**Issue by pearcec** [terraform-providers/terraform-provider-azurerm#10454]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
terraform0.14.exe -v                   
Terraform v0.14.0
+ provider registry.terraform.io/hashicorp/azurerm v2.45.1

Your version of Terraform is out of date! The latest version
is 0.14.5. You can update by downloading from https://www.terraform.io/downloads.html
```

### Affected Resource(s)

* `azurerm_key_vault_secret`

### Terraform Configuration Files

```hcl
provider "azurerm" {
    features {}
}

data "azurerm_client_config" "current" {}

resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "East US"
}

resource "azurerm_key_vault" "example" {
  name                       = "notyourkeyvaultvault"
  location                   = azurerm_resource_group.example.location
  resource_group_name        = azurerm_resource_group.example.name
  tenant_id                  = data.azurerm_client_config.current.tenant_id
  sku_name                   = "premium"
  soft_delete_retention_days = 7

  access_policy {
    tenant_id = data.azurerm_client_config.current.tenant_id
    object_id = data.azurerm_client_config.current.object_id

    key_permissions = [
      "create",
      "get",
    ]

    secret_permissions = [
      "set",
      "get",
      "delete",
      "purge",
      "recover"
    ]
  }
}

resource "azurerm_key_vault_secret" "example" {
  name         = "secret-sauce"
  value        = "szechuan"
  key_vault_id = azurerm_key_vault.example.id
}
```

### Debug Output

```
2021-02-03T13:46:33.817-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: {"recoveryId":"https://notyourkeyvaultvault.vault.azure.net/deletedsecrets/secret-sauce","deletedDate":1612377994,"scheduledPurgeDate":1612982794,"contentType":"","id":"https://notyourkeyvaultvault.vault.azure.net/secrets/secret-sauce/31508a7ebe214b1c81b812d8ef52ea67","attributes":{"enabled":true,"created":1612377962,"updated":1612377962,"recoveryLevel":"CustomizedRecoverable+Purgeable"},"tags":{}}
2021-02-03T13:46:33.817-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] Waiting for Secret "secret-sauce" (Key Vault "https://notyourkeyvaultvault.vault.azure.net/") to finish deleting..
2021-02-03T13:46:33.817-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] Waiting for state to become: [NotFound]
2021-02-03T13:46:33.817-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] AzureRM Request: 
2021-02-03T13:46:33.817-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: GET /secrets/secret-sauce/?api-version=2016-10-01 HTTP/1.1
2021-02-03T13:46:33.817-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Host: notyourkeyvaultvault.vault.azure.net
2021-02-03T13:46:33.817-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: User-Agent: Go/go1.15.5 (amd64-windows) go-autorest/v14.2.1 Azure-SDK-For-Go/v49.2.0 keyvault/2016-10-01 HashiCorp Terraform/0.14.0 (+https://www.terraform.io) Terraform Plugin SDK/1.13.1 terraform-provider-azurerm/2.45.1 pid-222c6c49-1b0a-5959-a213-6608f9eb8820
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] Deleted Secret "secret-sauce" (Key Vault "https://notyourkeyvaultvault.vault.azure.net/").
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] Purging Secret "secret-sauce" (Key Vault "https://notyourkeyvaultvault.vault.azure.net/")..
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] AzureRM Request: 
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: DELETE /deletedsecrets/secret-sauce?api-version=2016-10-01 HTTP/1.1
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Host: notyourkeyvaultvault.vault.azure.net
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: User-Agent: Go/go1.15.5 (amd64-windows) go-autorest/v14.2.1 Azure-SDK-For-Go/v49.2.0 keyvault/2016-10-01 HashiCorp Terraform/0.14.0 
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Accept-Encoding: gzip
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: 
2021-02-03T13:46:45.700-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: 
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] AzureRM Response for https://notyourkeyvaultvault.vault.azure.net/deletedsecrets/secret-sauce?api-version=2016-10-01: 
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: HTTP/2.0 401 Unauthorized
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Content-Length: 87
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Cache-Control: no-cache
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Content-Type: application/json; charset=utf-8
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Date: Wed, 03 Feb 2021 18:46:45 GMT
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Expires: -1
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Pragma: no-cache
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Strict-Transport-Security: max-age=31536000;includeSubDomains
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Www-Authenticate: Bearer authorization="https://login.windows.net/5ecec5bd-089d-4d8b-b2c1-5024715d6e42", resource="https://vault.azure.net"
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Content-Type-Options: nosniff
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Ms-Keyvault-Network-Info: conn_type=Ipv4;addr=152.208.0.197;act_addr_fam=InterNetwork;
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Ms-Keyvault-Region: eastus
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Ms-Keyvault-Service-Version: 1.2.139.0
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Powered-By: ASP.NET
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: 
2021-02-03T13:46:45.767-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: {"error":{"code":"Unauthorized","message":"Request is missing a Bearer or PoP token."}}
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] AzureRM Request: 
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: DELETE /deletedsecrets/secret-sauce?api-version=2016-10-01 HTTP/1.1
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Host: notyourkeyvaultvault.vault.azure.net
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: User-Agent: Go/go1.15.5 (amd64-windows) go-autorest/v14.2.1 Azure-SDK-For-Go/v49.2.0 keyvault/2016-10-01 HashiCorp Terraform/0.14.0 
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Ms-Correlation-Request-Id: 222b98a3-a9e8-3b63-bc58-d2ee22a902eb
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Accept-Encoding: gzip
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: 
2021-02-03T13:46:46.280-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: 
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: [DEBUG] AzureRM Response for https://notyourkeyvaultvault.vault.azure.net/deletedsecrets/secret-sauce?api-version=2016-10-01: 
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: HTTP/2.0 409 Conflict
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Content-Length: 121
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Cache-Control: no-cache
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Content-Type: application/json; charset=utf-8
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Date: Wed, 03 Feb 2021 18:46:47 GMT
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Expires: -1
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Pragma: no-cache
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: Strict-Transport-Security: max-age=31536000;includeSubDomains
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Content-Type-Options: nosniff
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Ms-Keyvault-Network-Info: conn_type=Ipv4;addr=152.208.0.197;act_addr_fam=InterNetwork;
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Ms-Keyvault-Region: eastus
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Ms-Keyvault-Service-Version: 1.2.139.0
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: X-Powered-By: ASP.NET
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: 
2021-02-03T13:46:46.357-0500 [DEBUG] plugin.terraform-provider-azurerm_v2.45.1_x5.exe: {"error":{"code":"Conflict","message":"Secret is currently being deleted.","innererror":{"code":"ObjectIsBeingDeleted"}}}
2021/02/03 13:46:46 [DEBUG] azurerm_key_vault_secret.example: apply errored, but we're indicating that via the Error pointer rather than returning it: purging Secret "secret-sauce" (Key Vault "https://notyourkeyvaultvault.vault.azure.net/"): keyvault.BaseClient#PurgeDeletedSecret: Failure responding to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code="Conflict" Message="Secret is currently being deleted." InnerError={"code":"ObjectIsBeingDeleted"}
```

### Expected Behaviour

The secret should be purged without conflict

### Actual Behaviour

```
C:\Users\pearcec\bin\terraform0.14.exe destroy -auto-approve
azurerm_key_vault_secret.example: Destroying... [id=https://notyourkeyvaultvault.vault.azure.net/secrets/secret-sauce/31508a7ebe214b1c81b812d8ef52ea67]
azurerm_key_vault_secret.example: Still destroying... [id=https://notyourkeyvaultvault.vault.azur...sauce/31508a7ebe214b1c81b812d8e2ea67, 10s elapsed]

Error: purging Secret "secret-sauce" (Key Vault "https://notyourkeyvaultvault.vault.azure.net/"): keyvault.BaseClient#PurgeDeletedSecret: Failure responding to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code="Conflict" Message="Secret is currently being deleted." InnerError={"code":"ObjectIsBeingDeleted"}
```

### Steps to Reproduce

1. `terraform apply`
2. `terraform destroy`

### Important Factoids

* If I run in a couple seconds later it deletes just fine.
* It definitely reaches the state NotFound

### References
* https://github.com/terraform-providers/terraform-provider-azurerm/blob/82b73af7a3319e3abb2bc1eb083be86941e9e9ab/azurerm/internal/services/keyvault/internal.go#L41
* #5659
* #9911
* #9984

**Comment by dkirrane** [comment]
Hit the same issue, had to re-run terraform destroy as a workaround. It doesn't fail the second time.

**Comment by pearcec** [comment]
Started to draft up a fix https://github.com/terraform-providers/terraform-provider-azurerm/pull/10577

**Comment by kemmemha** [comment]
Thanks for the effort you put into solving the issue @pearcec ! This issue corrupts our tf state in some cases so a simple re-run does only work in around 50% of the cases. We are looking forward to the fix 👍

**Comment by vizel014** [comment]
Thank you, @pearcec 
Also waiting for a solution, it just does not let perform terraform destroy,
Is it going to go into v2.48.0?

Error: purging Secret "****" (Key Vault ****"): keyvault.BaseClient#PurgeDeletedSecret: Failure responding to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code="Conflict" Message="Secret is currently being deleted." InnerError={"code":"ObjectIsBeingDeleted"}

**Comment by jason-johnson** [comment]
> Thanks for the effort you put into solving the issue @pearcec ! This issue corrupts our tf state in some cases so a simple re-run does only work in around 50% of the cases. We are looking forward to the fix 👍

In our case, the failure is 100% because the first run successfully deletes the key vault policies so the second run does not have permissions on the key vaults anymore.  So we're also looking very forward to the fix. :)

**Comment by jaswope** [comment]
One potential workaround that can work if you don't _need_ to purge the KeyVault entries (for example, if you aren't recreating the entries with the same name in the same KeyVault), is to disable the purge-on-destroy functionality until this is fixed.

```terraform
provider "azurerm" {
    features {
        key_vault {
            purge_soft_delete_on_destroy = false
        }
    }
}
```

**Comment by jason-johnson** [comment]
That's not an option for us.  We actually create environments for certain "feature" branches so users can make changes to the infrastructure without impacting others and these are destroyed when the feature branch is merged back in.  But we have to re-use the names for these feature branches because they get a dns name which is registered with our IDP.

**Comment by tombuildsstuff** [comment]
Fixed via #10577

**Comment by jason-johnson** [comment]
@tombuildsstuff  Hi, thanks for the fix.  This seems to not be happening anymore but now we get a timeout failure trying to delete a key vault policy.
